pkg/bucket/bucket_controller.go

```diff
@@ -0,0 +1,196 @@
+	bl := &BucketListener{
+		provisionerName:   provisionerName,
+		provisionerClient: client,
+	}
+
+	return bl
+}
+
+// Add attempts to create a bucket for a given bucket. This function must be idempotent
+// Return values
+//    nil - Bucket successfully provisioned
+//    non-nil err - Internal error                                [requeue'd with exponential backoff]
+func (b *BucketListener) Add(ctx context.Context, bucket *v1alpha1.Bucket) error {
+	klog.V(3).InfoS("Add Bucket",
+		"name", bucket.Name,
+		"bucketclass", bucket.Spec.BucketClassName,
+	)
+
+	if !strings.EqualFold(bucket.Spec.Provisioner, b.provisionerName) {
+		klog.V(5).InfoS("Skipping bucket for provisiner",
+			"bucket", bucket.Name,
+			"provisioner", bucket.Spec.Provisioner,
+		)
+		return nil
+	}
+
+	if bucket.Status.BucketAvailable {
+		klog.V(5).InfoS("BucketExists",
+			"bucket", bucket.Name,
+			"provisioner", bucket.Spec.Provisioner,
+		)
+		return nil
+	}
+
+	proto, err := bucket.Spec.Protocol.ConvertToExternal()
+	if err != nil {
+		klog.ErrorS(err, "Invalid protocol",
+			"bucket", bucket.Name)
+
+		return errors.Wrap(err, "Failed to parse protocol for API")
+	}
+
+	req := &cosi.ProvisionerCreateBucketRequest{
+		Parameters: bucket.Spec.Parameters,
+		Protocol:   proto,
+		Name:       bucket.Name,
+	}
+
+	rsp, err := b.provisionerClient.ProvisionerCreateBucket(ctx, req)
+	if err != nil {
+		if status.Code(err) != codes.AlreadyExists {
+			klog.ErrorS(err, "Failed to create bucket",
+				"bucket", bucket.Name)
+			return errors.Wrap(err, "Failed to create bucket")
+		}
+	}
+
+	bucket.Spec.BucketID = rsp.BucketId
+	bucket.Status.Message = "Bucket Provisioned"
+	bucket.Status.BucketAvailable = true
+
+	// if this step fails, then controller will retry with backoff
+	if _, err := b.Buckets().Update(ctx, bucket, metav1.UpdateOptions{}); err != nil {
+		klog.ErrorS(err, "Failed to update bucket",
+			"bucket", bucket.Name)
+		return errors.Wrap(err, "Failed to update bucket")
+	}
+
+	return nil
+}
+
+// Update attempts to reconcile changes to a given bucket. This function must be idempotent
+// Return values
+//    nil - Bucket successfully reconciled
+//    non-nil err - Internal error                                [requeue'd with exponential backoff]
+func (b *BucketListener) Update(ctx context.Context, old, new *v1alpha1.Bucket) error {
+	klog.V(3).InfoS("Update Bucket",
+		"name", old.Name)
+
+	return nil
+}
+
+// Delete attemps to delete a bucket. This function must be idempotent
+// Return values
+//    nil - Bucket successfully deleted
+//    non-nil err - Internal error                                [requeue'd with exponential backoff]
+func (b *BucketListener) Delete(ctx context.Context, bucket *v1alpha1.Bucket) error {
+	klog.V(3).InfoS("Delete Bucket",
+		"name", bucket.Name,
+		"bucketclass", bucket.Spec.BucketClassName,
+	)
+
+	if !strings.EqualFold(bucket.Spec.Provisioner, b.provisionerName) {
```

Comment:
Is there a common place in the code path where this test can be done before any CRUD methods are called?

Reply:
Its not that simple - underlying framework code cannot do that because in case of things like BAs, we have to trasitively find the bucket before inferring the provisioner